drivers/flash/flash_ll_nrfxlib.c

```diff
@@ -0,0 +1,337 @@
+
+#include <logging/log.h>
+#define LOG_MODULE_NAME blectrl_flash
+LOG_MODULE_REGISTER(LOG_MODULE_NAME);
+
+#define FLASH_DRIVER_WRITE_BLOCK_SIZE 1
+
+static struct {
+	/** Used to ensure a single ongoing operation at any time.  */
+	struct k_sem sem;
+	const void *data;
+	off_t addr;
+	u16_t len;
+	u16_t prev_len;
+	u32_t tmp_word;      /**< Used for unalinged writes. */
+	/* NOTE: Read is not async, so not a part of this enum. */
+	enum {
+		FLASH_OP_NONE,
+		FLASH_OP_WRITE,
+		FLASH_OP_ERASE
+	} op;
+} flash_state;
+
+/* Forward declarations */
+static int btctlr_flash_read(struct device *dev,
+			     off_t offset,
+			     void *data,
+			     size_t len);
+static int btctlr_flash_write(struct device *dev,
+			      off_t offset,
+			      const void *data,
+			      size_t len);
+static int btctlr_flash_erase(struct device *dev,
+			      off_t offset,
+			      size_t size);
+static int btctlr_flash_write_protection_set(struct device *dev,
+					     bool enable);
+
+static int flash_op_execute(void);
+
+#if defined(CONFIG_FLASH_PAGE_LAYOUT)
+static void btctlr_flash_page_layout_get(
+	struct device *dev,
+	const struct flash_pages_layout **layout,
+	size_t *layout_size);
+#endif /* defined(CONFIG_FLASH_PAGE_LAYOUT) */
+
+static const struct flash_driver_api btctrl_flash_api = {
+	.read = btctlr_flash_read,
+	.write = btctlr_flash_write,
+	.erase = btctlr_flash_erase,
+	.write_protection = btctlr_flash_write_protection_set,
+#if defined(CONFIG_FLASH_PAGE_LAYOUT)
+	.page_layout = btctlr_flash_page_layout_get,
+#endif  /* CONFIG_FLASH_PAGE_LAYOUT */
+	.write_block_size = FLASH_DRIVER_WRITE_BLOCK_SIZE
+};
+
+
+/* Utility functions */
+static inline bool is_addr_valid(off_t addr, size_t len)
+{
+	if ((addr + len) > NRF_FICR->CODEPAGESIZE * NRF_FICR->CODESIZE ||
+	    addr < 0) {
+		return false;
+	}
+
+	return true;
+}
+
+static inline bool is_aligned_32(off_t addr)
+{
+	return ((addr & 0x3) == 0);
+}
+
+static inline off_t align_32(off_t addr)
+{
+	return (addr & ~0x3);
+}
+
+static inline size_t bytes_to_words(size_t bytes)
+{
+	return bytes / sizeof(u32_t);
+}
+
+static inline bool is_page_aligned(off_t addr)
+{
+	return (addr & (NRF_FICR->CODEPAGESIZE - 1)) == 0;
+}
+
+static void flash_operation_complete_callback(u32_t status)
+{
+	__ASSERT_NO_MSG(flash_state.op == FLASH_OP_WRITE ||
+			flash_state.op == FLASH_OP_ERASE);
+
+	int err;
+
+	flash_state.addr += flash_state.prev_len;
+	flash_state.data = (const void *) ((intptr_t) flash_state.data
+					   + flash_state.prev_len);
+	flash_state.len -= flash_state.prev_len;
+
+	if (flash_state.len > 0) {
+		err = flash_op_execute();
+		/* All inputs should have been validated on the first call. */
+		__ASSERT(err == 0, "Continued flash operation failed\n");
+	} else {
+		flash_state.op = FLASH_OP_NONE;
+		k_sem_give(&flash_state.sem);
+	}
+}
+
+static size_t offset_32(const void *addr)
+{
+	return (const u32_t) addr & 0x3;
+}
+
+static size_t unaligned_word_copy(u32_t *word_dst,
```

Comment:
unaligned_truncated_word_copy ?

Reply:
I don't think that really helps. Are the docs I proposed above good enough?

Reply:
The docs are good enough